src/mds/Server.cc

```diff
@@ -5061,15 +5115,86 @@ void Server::handle_client_setattr(MDRequestRef& mdr)
+      // The xlock(filelock) has been dropped, reset the objver
+      mdr->objver = ULONG_MAX;
       return;
     }
+
+    if (truncating_smaller && req->get_data().length()) {
+      struct ceph_fscrypt_last_block_header header;
+      memset(&header, 0, sizeof(header));
+      auto bl = req->get_data().cbegin();
+      bufferlist *data = new bufferlist();
+      DECODE_START(1, bl);
+      decode(header.objver, bl);
+      decode(header.file_offset, bl);
+      decode(header.block_size, bl);
+      DECODE_FINISH(bl);
+
+      dout(20) << __func__ << " mdr->objver:" << mdr->objver
+               << " mdr->retry:" << mdr->retry
+               << " header.objver: " << header.objver
+               << " header.file_offset: " << header.file_offset
+               << " header.block_size: " << header.block_size
+               << dendl;
+
+      // Always read Rados for the first time just after acquiring
+      // the xclock for the filelock to make sure that no client has
+      // changed the object during this time.
+      //
+      // To make sure that the objvers are the same or let the client
+      // retry the size trucate request.
+      if (mdr->objver != ULONG_MAX) {
+        if (mdr->objver != header.objver) {
+          dout(5) << __func__ << ": header.objver:" << header.objver
+                  << " != current objver:" << mdr->objver
+                  << ", let client retry it!" << dendl;
+          respond_to_request(mdr, -CEPHFS_EAGAIN);
+	  return;
+        }
+        // Reset the objver just in case later the request will be
+        // queued to retry by dropping the xlock(filelock).
+        mdr->objver = ULONG_MAX;
+      } else {
+        auto layout = pip->layout;
+        /*
+         * Try to get the object version after acquiring the xlock
+         * for filelock. During this it won't drop the locks when
+         * trying this to ban any client to update this object in
+         * Rados.
+         *
+         * The extents.size() will be 1, because kclient will forbid
+         * the fscrypt block size to be larger than an object's size.
+         * And also one block won't cross two different objects.
+         */
+        std::vector<ObjectExtent> extents;
+        Striper::file_to_extents(g_ceph_context, cur->ino(), &layout, header.file_offset,
+                                 header.block_size, pip->truncate_size, extents);
+	ceph_assert(extents.size() == 1);
+        mds->objecter->read_trunc(extents[0].oid, object_locator_t(layout.pool_id),
+                                  header.file_offset, (uint64_t)8, mdr->snapid, data,
+                                  0, pip->truncate_size, pip->truncate_seq,
+                                  new C_OnFinisher(new C_IO_MDC_ReadtruncFinish(mdcache, cur, mdr, data),
+                                                   mds->finisher),
+                                  &mdr->objver);
+        return;
```

Comment:
This still doesn't look right to me, as I don't see how you're handling the case where the last block is inside an existing hole in the file.

IOW, suppose the client reads the last block for a truncate and and finds that the backing object does not exist. The client sends a 0 object version in the setattr request, along with an encrypted last block. However, before that setattr request arrives at the MDS another client issues a write and creates the object and writes some data to it.

What happens at that point? I don't see where you're ensuring that the trailing object still no longer exists when you go to issue the truncate and write.

Reply:
> This still doesn't look right to me, as I don't see how you're handling the case where the last block is inside an existing hole in the file.
> 
> IOW, suppose the client reads the last block for a truncate and and finds that the backing object does not exist. The client sends a 0 object version in the setattr request, along with an encrypted last block. However, before that setattr request arrives at the MDS another client issues a write and creates the object and writes some data to it.
> 
> What happens at that point? I don't see where you're ensuring that the trailing object still no longer exists when you go to issue the truncate and write.

In Line#5163 above will ensure this. In that case the objvers won't be same and kclient will set the `header.objver = 0`, and in MDS the new read `mdr->objver > 0`. Will let kclient retry it.

NOTE: The `mdr->objver` will be initialized to `ULONG_MAX` and will always read that object to get the `objver` just after xlock succeeds.

The logic here is:

1, kclient will read that object and if it's a hole, then just set the `header.objver = 0`.
2, kclient sends the setattr request to MDS.
3, in MDS it will try to `xlock` the `in->filelock` first, which will revoke the `Fsxrw` caps from all the clients, so no client could write new contents to the file.
4, in MDS it will try to read that object just after `xlock` succeeds, if the new read `objver` is changed. No matter it becomes a hole or written new contents to it. The MDS will fail the request and let kclient to retry it.
5, if the `objver` doesn't change since Step1, then we can continue.
6, journal the request, early reply the request
7, flush the journal
8, apply the change
9, drop the `xlock(in->filelock)` and safe reply the request